In [2]:
import pandas as pd
import numpy as np

In [2]:
fcm = pd.read_csv('data/final_marked.csv')
fcm.head()

patent_number  priority_date cite_subclass focal_subclass  first_seen
0       D466542           1902       D16/311        D16/303           1
1       D466542           1902         2/428        D16/303           1
2       D466542           1902         2/452        D16/303           1
3       D466542           1902         2/442        D16/303           1
4       D466542           1902         2/445        D16/303           1

In [5]:
master = pd.read_csv('data/master.csv')
master.head()

patent_number  grant_year  app_year  num_inventors  us_inventor  \
0       D257752        1981      1980              1          1.0   
1       D257924        1981      1980              1          1.0   
2       D258382        1981      1980              2          1.0   
3       D258383        1981      1980              2          1.0   
4       D258571        1981      1980              1          1.0   

   cite_foreign_patent  is_missing  num_design_cited  num_utility_cited  \
0                  0.0           1                 1                  2   
1                  0.0           1                 1                  2   
2                  0.0           1                 2                  1   
3                  0.0           1                 2                  1   
4                  0.0           1                 1                  1   

   non-pat_refs  num_figures  num_assignees  priority_date  
0           1.0          6.0              1           1980  
1           1.0          6.0              1           1980  
2           0.0          5.0              1           1980  
3           0.0          5.0              1           1980  
4           0.0          2.0              1           1980

In [47]:
master.shape

(525490, 13)

In [3]:
print(fcm.shape, fcm.patent_number.nunique())

(11447580, 5) 503128


Let's examine how many novel combinations exist

In [4]:
positive = fcm.loc[fcm['first_seen'] == 1]
positive.head()

patent_number  priority_date cite_subclass focal_subclass  first_seen
0       D466542           1902       D16/311        D16/303           1
1       D466542           1902         2/428        D16/303           1
2       D466542           1902         2/452        D16/303           1
3       D466542           1902         2/442        D16/303           1
4       D466542           1902         2/445        D16/303           1

In [5]:
positive.patent_number.nunique()

279931

In [48]:
positive.patent_number.nunique() / master.shape[0]

0.5327047136957886

Hm... 53.27% novel? lets see

Note: after veiwing a sample of the data, it indeed as been marked correctly.So this measure is accurate

Let's examin how many novel design(focal) -> design(reference) are made for each patent 

In [18]:
d2d = fcm.loc[(fcm['focal_subclass'].str.contains('D')) & fcm['cite_subclass'].str.contains('D')]
d2d.head()

patent_number  priority_date cite_subclass focal_subclass  first_seen
0        D466542           1902       D16/311        D16/303           1
5        D466542           1902       D16/303        D16/303           1
9        D499408           1904       D14/444        D14/444           1
10       D499408           1904       D14/441        D14/444           1
12       D510547           1905       D12/111        D12/111           1

In [33]:
num_d2d = d2d.groupby('patent_number')['first_seen'].apply(lambda x: np.count_nonzero(x))

In [34]:
d2u = fcm.loc[(fcm['focal_subclass'].str.contains('D')) & (~fcm['cite_subclass'].str.contains('D'))]
d2u.head()

patent_number  priority_date cite_subclass focal_subclass  first_seen
1       D466542           1902         2/428        D16/303           1
2       D466542           1902         2/452        D16/303           1
3       D466542           1902         2/442        D16/303           1
4       D466542           1902         2/445        D16/303           1
6       D499408           1904     312/223.2        D14/444           1

In [36]:
num_d2u = d2u.groupby('patent_number')['first_seen'].apply(lambda x: np.count_nonzero(x))

In [37]:
fcm.loc[(~fcm['focal_subclass'].str.contains('D')) & (~fcm['cite_subclass'].str.contains('D'))]

patent_number  priority_date cite_subclass focal_subclass  first_seen
1758           D264116           1979        138/99         405/43           1
1759           D264116           1979       138/111         405/43           1
1760           D264116           1979       138/112         405/43           1
1761           D264116           1979       138/115         405/43           1
1762           D264116           1979       138/178         405/43           1
1763           D264116           1979        405/48         405/43           1
1765           D264116           1979        405/43         405/43           1
1766           D264116           1979       210/455         405/43           1
3463           D260951           1980         5/668          5/675           1
3479           D260951           1980         5/680          5/675           1
3480           D260951           1980         5/679          5/675           1
3481           D260951           1980         5/678          5/675           1
3482           D260951           1980         5/499          5/675           1
3485           D260951           1980         5/915          5/675           1
3488           D260951           1980         5/679          5/915           1
3489           D260951           1980         5/678          5/915           1
3490           D260951           1980         5/499          5/915           1
3491           D260951           1980         5/668          5/915           1
3492           D260951           1980         5/915          5/915           1
3493           D260951           1980         5/683          5/915           1
3494           D260951           1980         5/682          5/915           1
3509           D260951           1980         5/680          5/915           1
3528           D260951           1980         5/682          5/675           1
3635           D261247           1980         63/13          63/13           1
3636           D261247           1980       63/29.1          63/13           1
3637           D261247           1980         63/31          63/13           1
3823           D260951           1980         5/683          5/675           1
4450           D262996           1980       206/382        206/366           1
4451           D262996           1980       206/350        206/366           1
4453           D262996           1980       206/370        206/380           1
...                ...            ...           ...            ...         ...
10342193       D729359           2014         4/619          4/656           1
10342194       D729359           2014         4/628          4/656           1
10342195       D729359           2014         4/630          4/656           1
10342196       D729359           2014         4/648          4/656           1
10342204       D729359           2014         4/650          4/656           1
10342264       D729359           2014       220/629          4/656           1
10342435       D729359           2014       220/485          4/656           1
10342436       D729359           2014         4/656          4/656           1
11192048       D773626           2015        62/298       62/259.1           1
11192052       D773626           2015        62/298        165/121           1
11192053       D773626           2015      62/259.1        165/121           1
11192070       D773626           2015        62/298         62/407           1
11192071       D773626           2015      62/259.1         62/407           1
11192074       D773626           2015      62/259.1       62/259.1           1
11362805       D785497           2015       47/66.6         47/62R           1
11364050       D785497           2015      211/59.2         47/62N           1
11364051       D785497           2015      211/59.4         47/62N           1
11364052       D785497           2015       221/102         47/62N           1
11364053       D785497           2015       221/303        

In [49]:
(positive.loc[(positive['focal_subclass'].str.contains('D')) & (positive['cite_subclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.3256008677615178

In [50]:
(positive.loc[(positive['focal_subclass'].str.contains('D')) & (~positive['cite_subclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.4104987725741689

In [52]:
(positive.loc[(~positive['focal_subclass'].str.contains('D')) & (~positive['cite_subclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.004961083940702963

In [53]:
(positive.loc[(~positive['focal_subclass'].str.contains('D')) & (positive['cite_subclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.004127576167006032

In [54]:
.41+.49

0.8999999999999999

In [3]:
fcm = pd.read_csv('data/final_main_marked.csv')
fcm.head()

patent_number  priority_date cite_mainclass focal_mainclass  first_seen
0       D466542           1902            D16             D16           1
1       D466542           1902              2             D16           1
2       D499408           1904            312             D14           1
3       D499408           1904            D14             D14           1
4       D510547           1905            D12             D12           1

In [4]:
print(fcm.shape, fcm.patent_number.nunique())

(2371580, 5) 503128


In [6]:
positive = fcm.loc[fcm['first_seen'] == 1]
positive.head()

patent_number  priority_date cite_mainclass focal_mainclass  first_seen
0       D466542           1902            D16             D16           1
1       D466542           1902              2             D16           1
2       D499408           1904            312             D14           1
3       D499408           1904            D14             D14           1
4       D510547           1905            D12             D12           1

In [7]:
positive.patent_number.nunique()

12460

In [9]:
positive.patent_number.nunique() / master.patent_number.nunique()

0.023711202877314506

When using mainclass, only 2.37% of data is novel. This seems more in line with previous data

In [11]:
(positive.loc[(positive['focal_mainclass'].str.contains('D')) & (positive['cite_mainclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.0023368665436069194

In [12]:
(positive.loc[(positive['focal_mainclass'].str.contains('D')) & (~positive['cite_mainclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.020276313535937885

In [13]:
(positive.loc[(~positive['focal_mainclass'].str.contains('D')) & (~positive['cite_mainclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.0025804487240480315

In [14]:
(positive.loc[(~positive['focal_mainclass'].str.contains('D')) & (positive['cite_mainclass'].str.contains('D'))]).patent_number.nunique() / master.shape[0]

0.00042436582998725

In [18]:
fcm.loc[(~fcm['focal_mainclass'].str.contains('D'))]

patent_number  priority_date cite_mainclass focal_mainclass  \
383           D264116           1979            138             405   
403           D264116           1979            405             405   
404           D264116           1979            210             405   
865           D274086           1979            D23              16   
887           D267356           1980            602              54   
1004          D260951           1980              5               5   
1057          D261247           1980            D11              63   
1058          D261247           1980             63              63   
1334          D262346           1980            D07             294   
1344          D262346           1980            D24             294   
1403          D262570           1980            D23             416   
1435          D262709           1980            D07              99   
1439          D262727           1980            D19             273   
1464          D262832           1980            D28             273   
1499          D262996           1980            206             206   
1510          D263055           1980            D15             184   
1511          D263055           1980            D34             184   
1512          D263055           1980            206             184   
1513          D263055           1980            220             184   
1514          D263055           1980            141             184   
1515          D263055           1980            184             184   
1517          D263055           1980            D15             220   
1518          D263055           1980            220             220   
1519          D263055           1980            141             220   
1520          D263055           1980            184             220   
1523          D263055           1980            D34             220   
1524          D263055           1980            206             220   
1654          D263298           1980            200             200   
1655          D263298           1980             70             200   
1656          D263298           1980            340             200   
...               ...            ...            ...             ...   
2022592       D713464           2013            D32               1   
2022593       D713464           2013            D30               1   
2022594       D713464           2013            D11               1   
2022596       D713464           2013            D07               1   
2022607       D713464           2013            D21               1   
2054375       D725768           2013            D24              24   
2113930       D752638           2013            D14              14   
2147871       D725246           2014            D07               4   
2147872       D725246           2014            D23               4   
2147873       D725246           2014            D32               4   
2150962       D727482           2014              4               4   
2150963       D727482           2014            604               4   
2150964       D727482           2014            600               4   
2150965       D727482           2014            D24               4   
2150966       D727482           2014            D23               4   
2153841       D729359           2014            D32               4   
2153842       D729359           2014              4               4   
2153843       D729359           2014            D23               4   
2153844       D729359           2014            220               4   
2153845       D729359           2014            D03               4   
2153846       D729359           2014            D07               4   
2321421       D773626           2015            D23             165   
2321422       D773626           2015             62             165   
2321423       D773626           2015            D23              62   
2321424       D773626           

In [30]:
classes = pd.read_csv('data/citations.csv')
classes.head()

/home/tiera/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


uspc_mainclass_id uspc_subclass_id patent_number
0               248          248/423       3930272
1               403          403/107       3930272
2                 5             5/11       3930272
3                 5            5/100       3930272
4                47         47/58.1R       3930335

In [31]:
classes = classes.dropna()
classes.sample(10)

uspc_mainclass_id uspc_subclass_id patent_number
984812                  2            2/228       5784723
138233                482          482/122       4249729
1299603               362         362/97.3       6439731
662291                 63           63/3.2       5195335
2348695               D02          D02/911       D599999
2333462               D14          D14/205       D587241
844356                473          473/158       5516108
594023                 52        52/748.11       5076037
372911                  5            5/237       4704751
2421320               D19           D19/84       D662140

Trying to test out international classification to see if its viable for use

In [36]:
international = pd.read_csv('data/ipcr.tsv', delimiter='\t')
international.head()

/home/tiera/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (4,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


uuid patent_id classification_level section ipc_class  \
0  0000hj3ytmy8g9l2qa5x1hta5   D409748                  NaN       D        24   
1  0000oxxw6v7dgyaqh4zgawa8r   7785997                    A       H         1   
2  0000rjeskx4ks9c0g8rrrewjc   8018027                    A       H         1   
3  0000t5tjrt2cnpe6kgci6cieu   8994735                    A       G         9   
4  0000tz39tuertu6rp06umhlq2   8777443                    A       F        21   

  subclass main_group subgroup symbol_position classification_value  \
0       04        NaN      NaN             NaN                  NaN   
1        L         21     3205               F                    I   
2        L         31      119               L                    I   
3        G          5      395               L                    I   
4        S          9       02               L                    I   

  classification_status classification_data_source action_date  \
0                   NaN                        NaN  0000-00-00   
1                     B                          H  2010-08-31   
2                     B                          H  2011-09-13   
3                     B                          H  2015-03-31   
4                     B                          H  2014-07-15   

  ipc_version_indicator  sequence  
0            0000-00-00         0  
1            2006-01-01         0  
2            2006-01-01         0  
3            2006-01-01         0  
4            2006-01-01         0

In [38]:
m = master.patent_number

In [42]:
international.loc[international['patent_id'].isin(m)]['patent_id'].nunique()

129497

In [43]:
master.patent_number.nunique()

525490

In [44]:
129497/525490

0.2464309501608023

In [56]:
cpc = pd.read_csv('data/cpc_current.tsv', delimiter='\t', usecols=['uuid','patent_id'])
cpc.dtypes

uuid         object
patent_id     int64
dtype: object

It seems as though patent_id is just utility patents, since the data types are just ints, instead of strings

In [57]:
classes = pd.read_csv('data/uspcs_level.csv')
classes.head()

uspc_mainclass_id uspc_subclass_id patent_number
0               D19           D19/75       D257752
1               D06          D06/573       D257924
2               D23          D23/214       D258382
3               D23          D23/214       D258383
4               D09          D09/560       D258571

In [59]:
classes.dropna(inplace=True)
classes.loc[~classes['uspc_mainclass_id'].str.match('D')]

uspc_mainclass_id uspc_subclass_id patent_number
143                    5            5/675       D260951
144                    5            5/915       D260951
208                   63            63/13       D261247
480                   56          56/16.2       D262289
481                   56         56/DIG10       D262289
490                    5           5/53.1       D262332
501                  294            294/7       D262346
509                  281           281/45       D262381
551                  441           441/59       D262477
604                  416            416/5       D262570
618                  172          172/355       D262596
655                   99          99/421V       D262709
656                   99          99/443R       D262709
657                   99           99/448       D262709
668                  273          273/150       D262727
694                  273          273/409       D262819
701                  273          273/150       D262832
733                  206          206/366       D262996
734                  206          206/380       D262996
747                   63            63/23       D263028
755                  184          184/106       D263055
756                  220          220/573       D263055
818                   62           62/298       D263239
819                   62           62/448       D263239
820                   62           62/449       D263239
822                   40           40/361       D263241
823                   40           40/362       D263241
824                   40           40/366       D263241
825                   40           40/367       D263241
831                  273          273/241       D263245
...                  ...              ...           ...
609323               293          293/102       D684508
609326               200        200/61.54       D684511
609327                74           74/552       D684511
609332               296        296/26.11       D684515
609333               296         296/57.1       D684515
610030               280          280/847       D685306
615097                 1              1/1       D691570
619580               137          137/359       D698011
619583               137          137/359       D698012
620696               137          137/359       D699817
625855                16           16/235       D710928
625856                16           16/309       D710928
626358                15         15/144.2       D712446
630917                24           24/112       D737429
631090               604          604/118       D739930
635860                 4            4/315       D698426
635861                 4            4/460       D698426
638321                24            24/40       D702592
646387                 1              1/1       D713464
654885                24           24/108       D725768
666525                14             14/4       D752638
673272                 4            4/656       D725246
674421                 4          4/114.1       D727482
674422                 4            4/454       D727482
675423                 4            4/656       D729359
722504               165          165/121       D773626
722505                62         62/259.1       D773626
722506                62           62/407       D773626
732676                47           47/62N       D785497
732677                47           47/62R       D785497

[5512 rows x 3 columns]